evm/src/witness/operation.rs

```diff
@@ -155,6 +159,74 @@ pub(crate) fn generate_keccak_general<F: Field>(
     Ok(())
 }
 
+pub(crate) fn generate_poseidon_general<F: RichField>(
+    state: &mut GenerationState<F>,
+    mut row: CpuColumnsView<F>,
+) -> Result<(), ProgramError> {
+    row.op.poseidon_general = F::ONE;
+    let [(context, _), (segment, log_in1), (base_virt, log_in2), (len, log_in3)] =
+        stack_pop_with_log_and_fill::<4, _>(state, &mut row)?;
+    let len = u256_to_usize(len)?;
+
+    let base_address = MemoryAddress::new_u256s(context, segment, base_virt)?;
+    let input = (0..len)
+        .map(|i| {
+            let address = MemoryAddress {
+                virt: base_address.virt.saturating_add(i),
+                ..base_address
+            };
+            let val = state.memory.get(address);
+            state.traces.push_memory(MemoryOp::new(
+                MemoryChannel::Code,
+                state.traces.clock(),
+                address,
+                MemoryOpKind::Read,
+                val.0[0].into(),
+            ));
+
+            val.0[0] as u32
+        })
+        .collect_vec();
+    log::debug!("Hashing {:?}", input);
```

Comment:
nit: Maybe change to "Poseidon hashing {:?}" to differentiate with `generate_keccak_general`.